### DISCUSSION TITLE
How to use `nprogress` package to show a loading top-bar on Vike?

### DISCUSSION THREAD

**Discussion by Blankeos** [vikejs/vike#1611] [Discussion] [answered]
The description is the title. I want to add a top-bar for loading. I'm currently on SolidJS though.

**Comment by phonzammi** [comment, marked as answer]
Hi @Blankeos, 
First, you need to install nprogress. Run `npm install nprogress` or `pnpm add nprogress`. Additionally, if you are using TypeScript, install the types with `npm install -D @types/nprogress` or `pnpm add -D @types/nprogress`.
Next, import the CSS by adding `import "nprogress/nprogress.css"` inside your `LayoutDefault.tsx` file.

To show the top-bar loading when navigating between pages, use it with the `+onPageTransitionStart` and `+onPageTransitionEnd` hooks.

For example:
```ts
// +onPageTransitionStart.ts
import type { OnPageTransitionStartAsync } from "vike/types";
import NProgress from "nprogress"

export const onPageTransitionStart: OnPageTransitionStartAsync = async () => {
  console.log("Page transition start");
  NProgress.start();
};
```

```ts
//+onPageTransitionEnd.ts
import type { OnPageTransitionEndAsync } from "vike/types";
import NProgress from "nprogress"

export const onPageTransitionEnd: OnPageTransitionEndAsync = async () => {
  console.log("Page transition end");
  NProgress.done();
  NProgress.remove();
};
```

**Reply by Blankeos** [reply]
Thanks for the quick response!

**Reply by phonzammi** [reply]
Happy to help

**Comment by Blankeos** [comment]
Managed to do it.

Repo: https://github.com/Blankeos/hono-trpc

Example above uses:
- Hono + tRPC (separate backend)
- Vike + Solid + Tanstack Query (Has SSR via dehydrate & hydrate) + Nprogress

https://github.com/vikejs/vike/assets/38070918/7af55e3d-0e09-4ab6-95d7-10adf4209dbf

I'll let you close this one. Let me know if you want examples of this added onto the docs or smth. (out of topic on `nprogress` but..) this roughly took me roughly 7 hours of just fiddling around lol (specifically for SSR). Solid has no docs on hydration so I sort of took inspiration from this [Vike React Query SSR Example](https://github.com/alternativniy/vite-ssr-with-react-query/).

**Reply by phonzammi** [reply]
Very nice, thanks for sharing

**Reply by brillout** [reply]
:sparkles: Neat, thanks, I added your example to the docs https://github.com/vikejs/vike/commit/140867cdbe6d3d4e5f8c433ce2805b8a8eff2bbf.
